Bring interrupting boundary event test back

**Description**

We removed the following test:

```

  @Test
  // https://github.com/camunda-cloud/zeebe/issues/6587
  public void shouldInterruptBoundaryEvent() {
    // given
    final ProcessBuilder process = Bpmn.createExecutableProcess(PROCESS_ID);

    process
        .eventSubProcess("event_sub_proc")
        .startEvent("event_sub_start")
        .interrupting(true)
        .message(b -> b.name(MSG_NAME).zeebeCorrelationKeyExpression("key"))
        .endEvent("event_sub_end");

    final BpmnModelInstance model =
        process
            .startEvent("start_proc")
            .sequenceFlowId("toParallel")
            .serviceTask("serviceTask", s -> s.zeebeJobType("task"))
            .boundaryEvent("boundary")
            .cancelActivity(true)
            .message(m -> m.name("msg").zeebeCorrelationKeyExpression("key"))
            .endEvent("end_proc")
            .done();

    engineRule.deployment().withXmlResource(model).deploy();

    final long processInstanceKey =
        engineRule
            .processInstance()
            .ofBpmnProcessId(PROCESS_ID)
            .withVariables(Map.of("key", 123, "key1", 123))
            .create();

    // when
    RecordingExporter.messageSubscriptionRecords(MessageSubscriptionIntent.CREATED)
        .withProcessInstanceKey(processInstanceKey)
        .withMessageName("msg")
        .await();

    engineRule
        .message()
        .withName("msg")
        .withCorrelationKey("123")
        .withVariables(Map.of("key", "123"))
        .publish();

    RecordingExporter.processInstanceRecords()
        .withElementId("serviceTask")
        .withProcessInstanceKey(processInstanceKey)
        .withIntent(ProcessInstanceIntent.ELEMENT_TERMINATING)
        .await();

    engineRule
        .message()
        .withName(MSG_NAME)
        .withCorrelationKey("123")
        .withVariables(Map.of("key", "123"))
        .publish();

    // then
    assertThat(
            RecordingExporter.processInstanceRecords()
                .withProcessInstanceKey(processInstanceKey)
                .limitToProcessInstanceCompleted())
        .extracting(r -> tuple(r.getValue().getBpmnElementType(), r.getIntent()))
        .containsSubsequence(
            tuple(BpmnElementType.EVENT_SUB_PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATED),
            tuple(BpmnElementType.EVENT_SUB_PROCESS, ProcessInstanceIntent.ELEMENT_COMPLETED),
            tuple(BpmnElementType.PROCESS, ProcessInstanceIntent.ELEMENT_COMPLETED));
  }
```

Because it was flaky. It was a regression test which we should add again, but we need to find a way how we can make sure that the boundary event is terminated. Due to the processor migration we now write the activation and complete command in a batch. This is makes it currently harder to write this test and made the test flaky.

Tried to enable this test again, but found a bug with concurrently triggering boundary and event sub process.

Related model:

![diagram_1](https://user-images.githubusercontent.com/2758593/117106607-75e08680-ad80-11eb-8edf-3db24be6b43c.png)

```java

  @Test
  // https://github.com/camunda-cloud/zeebe/issues/6587
  public void shouldInterruptBoundaryEvent() {
    // given
    final ProcessBuilder process = Bpmn.createExecutableProcess(PROCESS_ID);

    process
        .eventSubProcess("event_sub_proc")
        .startEvent("event_sub_start")
        .interrupting(true)
        .message(b -> b.name(MSG_NAME).zeebeCorrelationKeyExpression("key"))
        .endEvent("event_sub_end");

    final BpmnModelInstance model =
        process
            .startEvent("start_proc")
            .sequenceFlowId("toParallel")
            .serviceTask("serviceTask", s -> s.zeebeJobType("task"))
            .boundaryEvent("boundary")
            .cancelActivity(true)
            .message(m -> m.name("msg").zeebeCorrelationKeyExpression("key"))
            .endEvent("end_proc")
            .done();

    engineRule.deployment().withXmlResource(model).deploy();

    final long processInstanceKey =
        engineRule
            .processInstance()
            .ofBpmnProcessId(PROCESS_ID)
            .withVariables(Map.of("key", 123, "key1", 123))
            .create();

    // when
    RecordingExporter.messageSubscriptionRecords(MessageSubscriptionIntent.CREATED)
        .withProcessInstanceKey(processInstanceKey)
        .withMessageName("msg")
        .await();
    //
    //    engineRule
    //        .message()
    //        .withName("msg")
    //        .withCorrelationKey("123")
    //        .withVariables(Map.of("key", "123"))
    //        .publish();
    //
    //    RecordingExporter.processInstanceRecords()
    //        .withElementId("serviceTask")
    //        .withProcessInstanceKey(processInstanceKey)
    //        .withIntent(ProcessInstanceIntent.ELEMENT_TERMINATING)
    //        .await();
    //
    //    engineRule
    //        .message()
    //        .withName(MSG_NAME)
    //        .withCorrelationKey("123")
    //        .withVariables(Map.of("key", "123"))
    //        .publish();

    engineRule.writeRecords(
        RecordToWrite.command()
            .message(
                MessageIntent.PUBLISH,
                new MessageRecord().setName("msg").setCorrelationKey("123").setTimeToLive(0)),
        RecordToWrite.command()
            .message(
                MessageIntent.PUBLISH,
                new MessageRecord().setName(MSG_NAME).setCorrelationKey("123").setTimeToLive(0)));

    // then
    assertThat(
            RecordingExporter.processInstanceRecords()
                .withProcessInstanceKey(processInstanceKey)
                .limitToProcessInstanceCompleted())
        .extracting(r -> tuple(r.getValue().getBpmnElementType(), r.getIntent()))
        .containsSubsequence(
            tuple(BpmnElementType.EVENT_SUB_PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATED),
            tuple(BpmnElementType.EVENT_SUB_PROCESS, ProcessInstanceIntent.ELEMENT_COMPLETED),
            tuple(BpmnElementType.PROCESS, ProcessInstanceIntent.ELEMENT_COMPLETED));
  }
```

Writing both publish messages causes that the process instance to become stuck. It can't complete the boundary event since the scope is interrupted and the event sub process can't terminate the service task, since it is already terminated by the boundary event (terminate command is rejected). Might make sense to react on these rejections?! @saig0 


```
/usr/lib/jvm/java-11-openjdk/bin/java -ea -Didea.test.cyclic.buffer.size=1048576 -javaagent:/opt/intellij/idea-IU-203.6682.168/lib/idea_rt.jar=46439:/opt/intellij/idea-IU-203.6682.168/bin -Dfile.encoding=UTF-8 -classpath /opt/intellij/idea-IU-203.6682.168/lib/idea_rt.jar:/opt/intellij/idea-IU-203.6682.168/plugins/junit/lib/junit5-rt.jar:/opt/intellij/idea-IU-203.6682.168/plugins/junit/lib/junit-rt.jar:/home/zell/goPath/src/github.com/zeebe-io/zeebe/engine/target/test-classes:/home/zell/goPath/src/github.com/zeebe-io/zeebe/engine/target/classes:/home/zell/goPath/src/github.com/zeebe-io/zeebe/util/target/classes:/home/zell/.m2/repository/org/apache/logging/log4j/log4j-api/2.14.1/log4j-api-2.14.1.jar:/home/zell/.m2/repository/com/fasterxml/jackson/core/jackson-databind/2.12.3/jackson-databind-2.12.3.jar:/home/zell/.m2/repository/com/fasterxml/jackson/core/jackson-core/2.12.3/jackson-core-2.12.3.jar:/home/zell/.m2/repository/com/fasterxml/jackson/datatype/jackson-datatype-jsr310/2.12.3/jackson-datatype-jsr310-2.12.3.jar:/home/zell/.m2/repository/org/springframework/spring-core/5.3.6/spring-core-5.3.6.jar:/home/zell/.m2/repository/org/springframework/spring-jcl/5.3.6/spring-jcl-5.3.6.jar:/home/zell/.m2/repository/org/springframework/spring-context/5.3.6/spring-context-5.3.6.jar:/home/zell/.m2/repository/org/springframework/spring-aop/5.3.6/spring-aop-5.3.6.jar:/home/zell/.m2/repository/org/springframework/spring-expression/5.3.6/spring-expression-5.3.6.jar:/home/zell/.m2/repository/org/springframework/boot/spring-boot/2.4.5/spring-boot-2.4.5.jar:/home/zell/.m2/repository/org/springframework/boot/spring-boot-actuator/2.4.5/spring-boot-actuator-2.4.5.jar:/home/zell/.m2/repository/org/springframework/boot/spring-boot-autoconfigure/2.4.5/spring-boot-autoconfigure-2.4.5.jar:/home/zell/.m2/repository/org/springframework/boot/spring-boot-actuator-autoconfigure/2.4.5/spring-boot-actuator-autoconfigure-2.4.5.jar:/home/zell/.m2/repository/org/agrona/agrona/1.10.0/agrona-1.10.0.jar:/home/zell/.m2/repository/org/slf4j/slf4j-api/1.7.30/slf4j-api-1.7.30.jar:/home/zell/goPath/src/github.com/zeebe-io/zeebe/protocol/target/classes:/home/zell/goPath/src/github.com/zeebe-io/zeebe/zb-db/target/classes:/home/zell/.m2/repository/org/rocksdb/rocksdbjni/6.19.3/rocksdbjni-6.19.3.jar:/home/zell/goPath/src/github.com/zeebe-io/zeebe/protocol-impl/target/classes:/home/zell/.m2/repository/org/msgpack/jackson-dataformat-msgpack/0.8.22/jackson-dataformat-msgpack-0.8.22.jar:/home/zell/goPath/src/github.com/zeebe-io/zeebe/logstreams/target/classes:/home/zell/goPath/src/github.com/zeebe-io/zeebe/dispatcher/target/classes:/home/zell/.m2/repository/com/netflix/concurrency-limits/concurrency-limits-core/0.3.6/concurrency-limits-core-0.3.6.jar:/home/zell/goPath/src/github.com/zeebe-io/zeebe/atomix/cluster/target/classes:/home/zell/goPath/src/github.com/zeebe-io/zeebe/atomix/utils/target/classes:/home/zell/.m2/repository/com/esotericsoftware/minlog/1.3.1/minlog-1.3.1.jar:/home/zell/goPath/src/github.com/zeebe-io/zeebe/snapshot/target/classes:/home/zell/goPath/src/github.com/zeebe-io/zeebe/journal/target/classes:/home/zell/.m2/repository/io/netty/netty-transport/4.1.63.Final/netty-transport-4.1.63.Final.jar:/home/zell/.m2/repository/io/netty/netty-resolver/4.1.63.Final/netty-resolver-4.1.63.Final.jar:/home/zell/.m2/repository/io/netty/netty-codec/4.1.63.Final/netty-codec-4.1.63.Final.jar:/home/zell/.m2/repository/io/netty/netty-common/4.1.63.Final/netty-common-4.1.63.Final.jar:/home/zell/.m2/repository/io/netty/netty-transport-native-epoll/4.1.63.Final/netty-transport-native-epoll-4.1.63.Final-linux-x86_64.jar:/home/zell/.m2/repository/io/netty/netty-transport-native-unix-common/4.1.63.Final/netty-transport-native-unix-common-4.1.63.Final.jar:/home/zell/.m2/repository/org/apache/commons/commons-math3/3.6.1/commons-math3-3.6.1.jar:/home/zell/.m2/repository/com/esotericsoftware/kryo/5.1.1/kryo-5.1.1.jar:/home/zell/.m2/repository/com/esotericsoftware/reflectasm/1.11.9/reflectasm-1.11.9.jar:/home/zell/.m2/repository/io/netty/netty-buffer/4.1.63.Final/netty-buffer-4.1.63.Final.jar:/home/zell/.m2/repository/com/google/guava/guava/30.1.1-jre/guava-30.1.1-jre.jar:/home/zell/.m2/repository/com/google/guava/failureaccess/1.0.1/failureaccess-1.0.1.jar:/home/zell/.m2/repository/com/google/guava/listenablefuture/9999.0-empty-to-avoid-conflict-with-guava/listenablefuture-9999.0-empty-to-avoid-conflict-with-guava.jar:/home/zell/.m2/repository/org/checkerframework/checker-qual/3.8.0/checker-qual-3.8.0.jar:/home/zell/.m2/repository/com/google/j2objc/j2objc-annotations/1.3/j2objc-annotations-1.3.jar:/home/zell/.m2/repository/com/fasterxml/jackson/core/jackson-annotations/2.12.3/jackson-annotations-2.12.3.jar:/home/zell/goPath/src/github.com/zeebe-io/zeebe/msgpack-core/target/classes:/home/zell/goPath/src/github.com/zeebe-io/zeebe/expression-language/target/classes:/home/zell/.m2/repository/org/camunda/feel/feel-engine/1.13.2/feel-engine-1.13.2.jar:/home/zell/.m2/repository/com/lihaoyi/fastparse_2.13/2.3.1/fastparse_2.13-2.3.1.jar:/home/zell/.m2/repository/com/lihaoyi/sourcecode_2.13/0.2.3/sourcecode_2.13-0.2.3.jar:/home/zell/.m2/repository/com/lihaoyi/geny_2.13/0.6.5/geny_2.13-0.6.5.jar:/home/zell/.m2/repository/org/scala-lang/scala-library/2.13.5/scala-library-2.13.5.jar:/home/zell/goPath/src/github.com/zeebe-io/zeebe/msgpack-value/target/classes:/home/zell/goPath/src/github.com/zeebe-io/zeebe/bpmn-model/target/classes:/home/zell/.m2/repository/org/camunda/bpm/model/camunda-xml-model/7.15.0/camunda-xml-model-7.15.0.jar:/home/zell/.m2/repository/uk/co/real-logic/sbe-tool/1.22.0/sbe-tool-1.22.0.jar:/home/zell/.m2/repository/io/prometheus/simpleclient/0.10.0/simpleclient-0.10.0.jar:/home/zell/goPath/src/github.com/zeebe-io/zeebe/protocol-asserts/target/classes:/home/zell/.m2/repository/javax/annotation/javax.annotation-api/1.3.2/javax.annotation-api-1.3.2.jar:/home/zell/.m2/repository/junit/junit/4.13.2/junit-4.13.2.jar:/home/zell/.m2/repository/org/hamcrest/hamcrest-core/2.2/hamcrest-core-2.2.jar:/home/zell/.m2/repository/org/junit/jupiter/junit-jupiter-params/5.7.1/junit-jupiter-params-5.7.1.jar:/home/zell/.m2/repository/org/apiguardian/apiguardian-api/1.1.0/apiguardian-api-1.1.0.jar:/home/zell/.m2/repository/org/junit/jupiter/junit-jupiter-api/5.7.1/junit-jupiter-api-5.7.1.jar:/home/zell/.m2/repository/org/opentest4j/opentest4j/1.2.0/opentest4j-1.2.0.jar:/home/zell/.m2/repository/org/junit/platform/junit-platform-commons/1.7.1/junit-platform-commons-1.7.1.jar:/home/zell/.m2/repository/org/assertj/assertj-core/3.19.0/assertj-core-3.19.0.jar:/home/zell/.m2/repository/org/msgpack/msgpack-core/0.8.22/msgpack-core-0.8.22.jar:/home/zell/goPath/src/github.com/zeebe-io/zeebe/test-util/target/classes:/home/zell/goPath/src/github.com/zeebe-io/zeebe/exporter-api/target/classes:/home/zell/goPath/src/github.com/zeebe-io/zeebe/clients/java/target/classes:/home/zell/goPath/src/github.com/zeebe-io/zeebe/gateway-protocol-impl/target/classes:/home/zell/goPath/src/github.com/zeebe-io/zeebe/gateway-protocol/target/classes:/home/zell/.m2/repository/io/grpc/grpc-protobuf/1.37.0/grpc-protobuf-1.37.0.jar:/home/zell/.m2/repository/com/google/api/grpc/proto-google-common-protos/2.1.0/proto-google-common-protos-2.1.0.jar:/home/zell/.m2/repository/io/grpc/grpc-protobuf-lite/1.37.0/grpc-protobuf-lite-1.37.0.jar:/home/zell/.m2/repository/io/grpc/grpc-stub/1.37.0/grpc-stub-1.37.0.jar:/home/zell/.m2/repository/io/grpc/grpc-core/1.37.0/grpc-core-1.37.0.jar:/home/zell/.m2/repository/com/google/code/gson/gson/2.8.6/gson-2.8.6.jar:/home/zell/.m2/repository/com/google/android/annotations/4.1.1.4/annotations-4.1.1.4.jar:/home/zell/.m2/repository/io/perfmark/perfmark-api/0.23.0/perfmark-api-0.23.0.jar:/home/zell/.m2/repository/io/grpc/grpc-netty/1.37.0/grpc-netty-1.37.0.jar:/home/zell/.m2/repository/io/netty/netty-codec-http2/4.1.63.Final/netty-codec-http2-4.1.63.Final.jar:/home/zell/.m2/repository/io/netty/netty-codec-http/4.1.63.Final/netty-codec-http-4.1.63.Final.jar:/home/zell/.m2/repository/io/netty/netty-handler-proxy/4.1.63.Final/netty-handler-proxy-4.1.63.Final.jar:/home/zell/.m2/repository/io/netty/netty-codec-socks/4.1.63.Final/netty-codec-socks-4.1.63.Final.jar:/home/zell/.m2/repository/com/google/protobuf/protobuf-java/3.15.8/protobuf-java-3.15.8.jar:/home/zell/.m2/repository/io/netty/netty-tcnative-boringssl-static/2.0.39.Final/netty-tcnative-boringssl-static-2.0.39.Final.jar:/home/zell/.m2/repository/io/netty/netty-handler/4.1.63.Final/netty-handler-4.1.63.Final.jar:/home/zell/.m2/repository/com/fasterxml/jackson/dataformat/jackson-dataformat-yaml/2.12.3/jackson-dataformat-yaml-2.12.3.jar:/home/zell/.m2/repository/org/testcontainers/testcontainers/1.15.3/testcontainers-1.15.3.jar:/home/zell/.m2/repository/org/apache/commons/commons-compress/1.20/commons-compress-1.20.jar:/home/zell/.m2/repository/org/rnorth/duct-tape/duct-tape/1.0.8/duct-tape-1.0.8.jar:/home/zell/.m2/repository/org/rnorth/visible-assertions/visible-assertions/2.1.2/visible-assertions-2.1.2.jar:/home/zell/.m2/repository/net/java/dev/jna/jna/5.8.0/jna-5.8.0.jar:/home/zell/.m2/repository/com/github/docker-java/docker-java-api/3.2.8/docker-java-api-3.2.8.jar:/home/zell/.m2/repository/com/github/docker-java/docker-java-transport-zerodep/3.2.8/docker-java-transport-zerodep-3.2.8.jar:/home/zell/.m2/repository/com/github/docker-java/docker-java-transport/3.2.8/docker-java-transport-3.2.8.jar:/home/zell/.m2/repository/org/springframework/spring-beans/5.3.6/spring-beans-5.3.6.jar:/home/zell/.m2/repository/org/yaml/snakeyaml/1.28/snakeyaml-1.28.jar:/home/zell/.m2/repository/io/grpc/grpc-api/1.37.0/grpc-api-1.37.0.jar:/home/zell/.m2/repository/io/grpc/grpc-context/1.37.0/grpc-context-1.37.0.jar:/home/zell/.m2/repository/com/google/code/findbugs/jsr305/3.0.2/jsr305-3.0.2.jar:/home/zell/.m2/repository/com/google/errorprone/error_prone_annotations/2.6.0/error_prone_annotations-2.6.0.jar:/home/zell/.m2/repository/org/codehaus/mojo/animal-sniffer-annotations/1.20/animal-sniffer-annotations-1.20.jar:/home/zell/.m2/repository/org/mockito/mockito-core/3.9.0/mockito-core-3.9.0.jar:/home/zell/.m2/repository/net/bytebuddy/byte-buddy/1.10.20/byte-buddy-1.10.20.jar:/home/zell/.m2/repository/net/bytebuddy/byte-buddy-agent/1.10.20/byte-buddy-agent-1.10.20.jar:/home/zell/.m2/repository/org/objenesis/objenesis/3.2/objenesis-3.2.jar:/home/zell/.m2/repository/org/awaitility/awaitility/4.0.3/awaitility-4.0.3.jar:/home/zell/.m2/repository/org/hamcrest/hamcrest/2.2/hamcrest-2.2.jar:/home/zell/.m2/repository/org/apache/commons/commons-lang3/3.12.0/commons-lang3-3.12.0.jar:/home/zell/goPath/src/github.com/zeebe-io/zeebe/logstreams/target/test-classes:/home/zell/goPath/src/github.com/zeebe-io/zeebe/build-tools/target/classes:/home/zell/.m2/repository/org/apache/logging/log4j/log4j-core/2.14.1/log4j-core-2.14.1.jar:/home/zell/.m2/repository/org/apache/logging/log4j/log4j-slf4j-impl/2.14.1/log4j-slf4j-impl-2.14.1.jar com.intellij.rt.junit.JUnitStarter -ideVersion5 -junit4 io.camunda.zeebe.engine.processing.bpmn.subprocess.InterruptingEventSubprocessConcurrencyTest,shouldInterruptBoundaryEvent
08:55:45.293 [Broker-0-LogStream-1] DEBUG io.camunda.zeebe.logstreams - Configured log appender back pressure at partition 1 as AppenderVegasCfg{initialLimit=1024, maxConcurrency=32768, alphaLimit=0.7, betaLimit=0.95}. Window limiting is disabled
08:55:45.462 [Broker-0-StreamProcessor-1] DEBUG io.camunda.zeebe.logstreams - Recovering state of partition 1 from snapshot
08:55:45.565 [Broker-0-StreamProcessor-1] INFO  org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.el.impl.feel.MessagePackValueMapper@185da7fb)), function-provider: io.camunda.zeebe.el.impl.feel.FeelFunctionProvider@4c57663b, clock: io.camunda.zeebe.el.impl.ZeebeFeelEngineClock@cfbebd, configuration: Configuration(false)]
08:55:45.709 [Broker-0-StreamProcessor-1] INFO  io.camunda.zeebe.logstreams - Recovered state of partition 1 from snapshot at position -1
08:55:45.727 [Broker-0-StreamProcessor-1] INFO  org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.el.impl.feel.MessagePackValueMapper@13ff6952)), function-provider: io.camunda.zeebe.el.impl.feel.FeelFunctionProvider@508e534e, clock: io.camunda.zeebe.el.impl.ZeebeFeelEngineClock@193c6903, configuration: Configuration(false)]
08:55:45.742 [Broker-0-StreamProcessor-1] INFO  org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.el.impl.feel.MessagePackValueMapper@9bcc2f8)), function-provider: io.camunda.zeebe.el.impl.feel.FeelFunctionProvider@3624ed8, clock: io.camunda.zeebe.el.impl.ZeebeFeelEngineClock@617010a9, configuration: Configuration(false)]
08:55:46.477 [Broker-0-StreamProcessor-1] INFO  org.camunda.feel.FeelEngine - Engine created. [value-mapper: CompositeValueMapper(List(io.camunda.zeebe.el.impl.feel.MessagePackValueMapper@a07f6fa)), function-provider: io.camunda.zeebe.el.impl.feel.FeelFunctionProvider@6f7cca06, clock: io.camunda.zeebe.el.impl.ZeebeFeelEngineClock@6f741338, configuration: Configuration(false)]
08:55:51.791 [] INFO  io.camunda.zeebe.test - Test failed, following records were exported:
08:55:51.838 [] INFO  io.camunda.zeebe.test - Compact log representation:
--------
	['C'ommand/'E'event/'R'ejection] [valueType] [intent] - #[position]->#[source record position]  P[partitionId]K[key] - [summary of value]
	P9K999 - key; #999 - record position; "ID" element/process id; @"elementid"/[P9K999] - element with ID and key
	Keys are decomposed into partition id and per partition key (e.g. 2251799813685253 -> P1K005). If single partition, the partition is omitted.
	Long IDs are shortened (e.g. 'startEvent_5d56488e-0570-416c-ba2d-36d2a3acea78' -> 'star..acea78'
--------
C DEPLOYMENT         CREATE         - #01-> -1    -1 - 
E PROC               CREATED        - #02->#01 P1K01 - process.xml -> "proc" (version:1)
E DEPLOYMENT         CREATED        - #03->#01 P1K02 - process.xml
E DEPLOYMENT         FULLY_DISTR    - #04->#01 P1K02 - 
C PROC_INST_CREATION CREATE         - #05-> -1    -1 - new <process "proc"> with variables: {key1=123, key=123}
E VAR                CREATED        - #06->#05 P1K04 - key1->123 in <process [P1K03]>
E VAR                CREATED        - #07->#05 P1K05 - key->123 in <process [P1K03]>
C PROC_INST          ACTIVATE       - #08->#05 P1K03 - PROCESS "proc" in <process "proc"[P1K03]>
E PROC_INST_CREATION CREATED        - #09->#05 P1K06 - new <process "proc"> with variables: {key1=123, key=123}
E PROC_INST          ACTIVATING     - #10->#08 P1K03 - PROCESS "proc" in <process "proc"[P1K03]>
E PROC_MSG_SUB       CREATING       - #11->#08 P1K07 - "messageName" (inter.) correlationKey: 123 @[P1K03] in <process "proc"[P1K03]> (no vars)
E PROC_INST          ACTIVATED      - #12->#08 P1K03 - PROCESS "proc" in <process "proc"[P1K03]>
C PROC_INST          ACTIVATE       - #13->#08    -1 - START_EVENT "start_proc" in <process "proc"[P1K03]>
C MSG_SUB            CREATE         - #14-> -1    -1 - "messageName" (inter.) correlationKey: 123 @[P1K03] in <process "proc"[P1K03]> (no vars)
E PROC_INST          ACTIVATING     - #15->#13 P1K08 - START_EVENT "start_proc" in <process "proc"[P1K03]>
E PROC_INST          ACTIVATED      - #16->#13 P1K08 - START_EVENT "start_proc" in <process "proc"[P1K03]>
C PROC_INST          COMPLETE       - #17->#13 P1K08 - START_EVENT "start_proc" in <process "proc"[P1K03]>
E MSG_SUB            CREATED        - #18->#14 P1K09 - "messageName" (inter.) correlationKey: 123 @[P1K03] in <process "proc"[P1K03]> (no vars)
C PROC_MSG_SUB       CREATE         - #19-> -1    -1 - "messageName" (inter.) @[P1K03] in <process ?[P1K03]> (no vars)
E PROC_INST          COMPLETING     - #20->#17 P1K08 - START_EVENT "start_proc" in <process "proc"[P1K03]>
E PROC_INST          COMPLETED      - #21->#17 P1K08 - START_EVENT "start_proc" in <process "proc"[P1K03]>
E PROC_INST          SEQ_FLOW_TAKEN - #22->#17 P1K10 - SEQUENCE_FLOW "toParallel" in <process "proc"[P1K03]>
C PROC_INST          ACTIVATE       - #23->#17 P1K11 - SERVICE_TASK "serviceTask" in <process "proc"[P1K03]>
E PROC_MSG_SUB       CREATED        - #24->#19 P1K07 - "messageName" (inter.) correlationKey: 123 @[P1K03] in <process "proc"[P1K03]> (no vars)
E PROC_INST          ACTIVATING     - #25->#23 P1K11 - SERVICE_TASK "serviceTask" in <process "proc"[P1K03]>
E PROC_MSG_SUB       CREATING       - #26->#23 P1K12 - "msg" (inter.) correlationKey: 123 @[P1K11] in <process "proc"[P1K03]> (no vars)
E JOB                CREATED        - #27->#23 P1K13 - P1K13 "task" @"serviceTask"[P1K11] 3 retries, in <process "proc"[P1K03]> (no vars)
E PROC_INST          ACTIVATED      - #28->#23 P1K11 - SERVICE_TASK "serviceTask" in <process "proc"[P1K03]>
C MSG_SUB            CREATE         - #29-> -1    -1 - "msg" (inter.) correlationKey: 123 @[P1K11] in <process "proc"[P1K03]> (no vars)
E MSG_SUB            CREATED        - #30->#29 P1K14 - "msg" (inter.) correlationKey: 123 @[P1K11] in <process "proc"[P1K03]> (no vars)
C PROC_MSG_SUB       CREATE         - #31-> -1    -1 - "msg" (inter.) @[P1K11] in <process ?[P1K03]> (no vars)
E PROC_MSG_SUB       CREATED        - #32->#31 P1K12 - "msg" (inter.) correlationKey: 123 @[P1K11] in <process "proc"[P1K03]> (no vars)
C MSG                PUBLISH        - #33-> -1 P0K01 - "msg" correlationKey: 123 (no vars)
C MSG                PUBLISH        - #34-> -1 P0K01 - "messageName" correlationKey: 123 (no vars)
E MSG                PUBLISHED      - #35->#33 P1K15 - "msg" correlationKey: 123 (no vars)
E MSG_SUB            CORRELATING    - #36->#33 P1K14 - "msg" (inter.) correlationKey: 123 @[P1K11] in <process "proc"[P1K03]> (no vars)
E MSG                EXPIRED        - #37->#33 P1K15 - "msg" correlationKey: 123 (no vars)
C PROC_MSG_SUB       CORRELATE      - #38-> -1    -1 - "msg" (inter.) correlationKey: 123 @[P1K11] in <process "proc"[P1K03]> (no vars)
E MSG                PUBLISHED      - #39->#34 P1K16 - "messageName" correlationKey: 123 (no vars)
E MSG_SUB            CORRELATING    - #40->#34 P1K09 - "messageName" (inter.) correlationKey: 123 @[P1K03] in <process "proc"[P1K03]> (no vars)
E MSG                EXPIRED        - #41->#34 P1K16 - "messageName" correlationKey: 123 (no vars)
C PROC_MSG_SUB       CORRELATE      - #42-> -1    -1 - "messageName" (inter.) correlationKey: 123 @[P1K03] in <process "proc"[P1K03]> (no vars)
E PROC_MSG_SUB       CORRELATED     - #43->#38 P1K12 - "msg" (inter.) correlationKey: 123 @[P1K11] in <process "proc"[P1K03]> (no vars)
E PROC_EVNT          TRIGGERING     - #44->#38 P1K17 -  @"boundary"[P1K11] in <process P1K01[P1K03]> (no vars)
C PROC_INST          TERMINATE      - #45->#38 P1K11 - SERVICE_TASK "serviceTask" in <process "proc"[P1K03]>
C MSG_SUB            CORRELATE      - #46-> -1    -1 - "msg" (inter.) @[P1K11] in <process "proc"[P1K03]> (no vars)
C MSG_SUB            CORRELATE      - #47-> -1    -1 - "messageName" (inter.) @[P1K03] in <process "proc"[P1K03]> (no vars)
E PROC_MSG_SUB       CORRELATED     - #48->#42 P1K07 - "messageName" (inter.) correlationKey: 123 @[P1K03] in <process "proc"[P1K03]> (no vars)
E PROC_EVNT          TRIGGERING     - #49->#42 P1K18 -  @"event_sub_start"[P1K03] in <process P1K01[P1K03]> (no vars)
C PROC_INST          TERMINATE      - #50->#42 P1K11 - SERVICE_TASK "serviceTask" in <process "proc"[P1K03]>
E PROC_INST          TERMINATING    - #51->#45 P1K11 - SERVICE_TASK "serviceTask" in <process "proc"[P1K03]>
C JOB                CANCEL         - #52->#45 P1K13 - P1K13 "task" @"serviceTask"[P1K11] 3 retries, in <process "proc"[P1K03]> (no vars)
E PROC_INST          TERMINATED     - #53->#45 P1K11 - SERVICE_TASK "serviceTask" in <process "proc"[P1K03]>
E PROC_EVNT          TRIGGERED      - #54->#45 P1K17 -  @"boundary"[P1K11] in <process P1K01[P1K03]> (no vars)
E PROC_INST          ACTIVATING     - #55->#45 P1K19 - BOUNDARY_EVENT "boundary" in <process "proc"[P1K03]>
E PROC_INST          ACTIVATED      - #56->#45 P1K19 - BOUNDARY_EVENT "boundary" in <process "proc"[P1K03]>
C PROC_INST          COMPLETE       - #57->#45 P1K19 - BOUNDARY_EVENT "boundary" in <process "proc"[P1K03]>
E MSG_SUB            CORRELATED     - #58->#46 P1K14 - "msg" (inter.) correlationKey: 123 @[P1K11] in <process "proc"[P1K03]> (no vars)
E MSG_SUB            CORRELATED     - #59->#47 P1K09 - "messageName" (inter.) correlationKey: 123 @[P1K03] in <process "proc"[P1K03]> (no vars)
R PROC_INST          TERMINATE      - #60->#50 P1K11 - SERVICE_TASK "serviceTask" in <process "proc"[P1K03]> !INVALID_STATE (Expected element instance with key '2251799813685259' to be present in state but not found.)
E JOB                CANCELED       - #61->#52 P1K13 - P1K13 "task" @"serviceTask"[P1K11] 3 retries, in <process "proc"[P1K03]> (no vars)
R PROC_INST          COMPLETE       - #62->#57 P1K19 - BOUNDARY_EVENT "boundary" in <process "proc"[P1K03]> !INVALID_STATE (Expected flow scope instance to be not interrupted but was interrupted by an event with id 'event_sub_proc'.)

-------------- Deployed Processes ----------------------
process.xml -> "proc" (version:1)[P1K01] ------
<?xml version="1.0" encoding="UTF-8" standalone="no"?>
<definitions xmlns:bpmndi="http://www.omg.org/spec/BPMN/20100524/DI" xmlns:dc="http://www.omg.org/spec/DD/20100524/DC" xmlns:di="http://www.omg.org/spec/DD/20100524/DI" xmlns:ns0="http://camunda.org/schema/zeebe/1.0" id="definitions_889d40c1-faee-4e2e-9816-3bdb97dd8f0e" targetNamespace="http://www.omg.org/spec/BPMN/20100524/MODEL" xmlns="http://www.omg.org/spec/BPMN/20100524/MODEL">
  <process id="proc" isExecutable="true">
    <subProcess id="event_sub_proc" name="event_sub_proc" triggeredByEvent="true">
      <startEvent id="event_sub_start" isInterrupting="true" name="event_sub_start">
        <outgoing>sequenceFlow_2de9a5ed-f2e9-4ecf-a747-fded93ed73df</outgoing>
        <messageEventDefinition id="messageEventDefinition_6e53c7e2-92a2-4c97-9750-6faf9080b5f2" messageRef="message_695f56ef-f4d3-4e7e-a9d5-1a42b7a524d1"/>
      </startEvent>
      <endEvent id="event_sub_end" name="event_sub_end">
        <incoming>sequenceFlow_2de9a5ed-f2e9-4ecf-a747-fded93ed73df</incoming>
      </endEvent>
      <sequenceFlow id="sequenceFlow_2de9a5ed-f2e9-4ecf-a747-fded93ed73df" sourceRef="event_sub_start" targetRef="event_sub_end"/>
    </subProcess>
    <startEvent id="start_proc" name="start_proc">
      <outgoing>toParallel</outgoing>
    </startEvent>
    <sequenceFlow id="toParallel" sourceRef="start_proc" targetRef="serviceTask"/>
    <serviceTask id="serviceTask" name="serviceTask">
      <extensionElements>
        <ns0:taskDefinition type="task"/>
      </extensionElements>
      <incoming>toParallel</incoming>
    </serviceTask>
    <boundaryEvent attachedToRef="serviceTask" cancelActivity="true" id="boundary" name="boundary">
      <outgoing>sequenceFlow_72e8e9ee-7c9e-4727-8508-36abd520fff7</outgoing>
      <messageEventDefinition id="messageEventDefinition_7cfc95b4-d1b4-4e10-8aa1-133ce8f99f12" messageRef="message_a0733087-f80c-4b54-a4ad-ce5914102c10"/>
    </boundaryEvent>
    <endEvent id="end_proc" name="end_proc">
      <incoming>sequenceFlow_72e8e9ee-7c9e-4727-8508-36abd520fff7</incoming>
    </endEvent>
    <sequenceFlow id="sequenceFlow_72e8e9ee-7c9e-4727-8508-36abd520fff7" sourceRef="boundary" targetRef="end_proc"/>
  </process>
  <message id="message_695f56ef-f4d3-4e7e-a9d5-1a42b7a524d1" name="messageName">
    <extensionElements>
      <ns0:subscription correlationKey="=key"/>
    </extensionElements>
  </message>
  <message id="message_a0733087-f80c-4b54-a4ad-ce5914102c10" name="msg">
    <extensionElements>
      <ns0:subscription correlationKey="=key"/>
    </extensionElements>
  </message>
  <bpmndi:BPMNDiagram id="BPMNDiagram_e94632ac-41d1-47a7-8f59-2fa1b27e3378">
    <bpmndi:BPMNPlane bpmnElement="proc" id="BPMNPlane_0ee40d89-ba0e-409c-84ea-16d17a999a86">
      <bpmndi:BPMNShape bpmnElement="event_sub_proc" id="BPMNShape_c6b28853-e3e0-466d-8fc8-5f916e91dcaa" isExpanded="true">
        <dc:Bounds height="200.0" width="350.0" x="100.0" y="250.0"/>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape bpmnElement="event_sub_start" id="BPMNShape_f0826b6b-cbb4-4db4-9405-45185eea316a">
        <dc:Bounds height="36.0" width="36.0" x="150.0" y="332.0"/>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape bpmnElement="event_sub_end" id="BPMNShape_ff72ce22-63d6-42bc-bffd-94512e42cb42">
        <dc:Bounds height="36.0" width="36.0" x="236.0" y="332.0"/>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge bpmnElement="sequenceFlow_2de9a5ed-f2e9-4ecf-a747-fded93ed73df" id="BPMNEdge_2bc236fe-011a-4bb5-b303-e5e459d2c0ad">
        <di:waypoint x="186.0" y="350.0"/>
        <di:waypoint x="236.0" y="350.0"/>
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape bpmnElement="start_proc" id="BPMNShape_3f78b22a-99a5-4661-9a86-6d7db68ac8bb">
        <dc:Bounds height="36.0" width="36.0" x="100.0" y="100.0"/>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape bpmnElement="serviceTask" id="BPMNShape_6e36494f-be10-4a79-b122-487fa5ae16ed">
        <dc:Bounds height="80.0" width="100.0" x="186.0" y="78.0"/>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge bpmnElement="toParallel" id="BPMNEdge_ed5c259a-ec16-448c-8ecd-fb2b002853cd">
        <di:waypoint x="136.0" y="118.0"/>
        <di:waypoint x="186.0" y="118.0"/>
      </bpmndi:BPMNEdge>
      <bpmndi:BPMNShape bpmnElement="boundary" id="BPMNShape_96d20dae-60d9-4df7-b90a-86471fe7fdbd">
        <dc:Bounds height="36.0" width="36.0" x="218.0" y="140.0"/>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNShape bpmnElement="end_proc" id="BPMNShape_10128415-4a09-4d21-9fe1-40a1021530d9">
        <dc:Bounds height="36.0" width="36.0" x="266.5" y="208.0"/>
      </bpmndi:BPMNShape>
      <bpmndi:BPMNEdge bpmnElement="sequenceFlow_72e8e9ee-7c9e-4727-8508-36abd520fff7" id="BPMNEdge_c87cb1cf-c4c3-46fd-83c1-3854e8373b08">
        <di:waypoint x="236.0" y="176.0"/>
        <di:waypoint x="236.0" y="226.0"/>
        <di:waypoint x="266.5" y="226.0"/>
      </bpmndi:BPMNEdge>
    </bpmndi:BPMNPlane>
  </bpmndi:BPMNDiagram>
</definitions>


--------------- Decomposed keys (for debugging) -----------------
   -1 <-> -1
P0K01 <-> 1
P1K01 <-> 2251799813685249
P1K02 <-> 2251799813685250
P1K03 <-> 2251799813685251
P1K04 <-> 2251799813685252
P1K05 <-> 2251799813685253
P1K06 <-> 2251799813685254
P1K07 <-> 2251799813685255
P1K08 <-> 2251799813685256
P1K09 <-> 2251799813685257
P1K10 <-> 2251799813685258
P1K11 <-> 2251799813685259
P1K12 <-> 2251799813685260
P1K13 <-> 2251799813685261
P1K14 <-> 2251799813685262
P1K15 <-> 2251799813685263
P1K16 <-> 2251799813685264
P1K17 <-> 2251799813685265
P1K18 <-> 2251799813685266
P1K19 <-> 2251799813685267

08:55:51.842 [] DEBUG io.camunda.zeebe.util.buffer - Close stream processor
08:55:51.844 [Broker-0-StreamProcessor-1] DEBUG io.camunda.zeebe.logstreams - Closed stream processor controller Broker-0-StreamProcessor-1.
08:55:51.876 [Broker-0-LogStream-1] INFO  io.camunda.zeebe.logstreams - Close appender for log stream stream-1
08:55:51.876 [stream-1-write-buffer] DEBUG io.camunda.zeebe.dispatcher - Dispatcher closed
08:55:51.877 [Broker-0-LogStream-1] INFO  io.camunda.zeebe.logstreams - On closing logstream stream-1 close 2 readers
08:55:51.877 [] DEBUG io.camunda.zeebe.broker.test - Clean up test files on path /tmp/junit129773564562340249
08:55:51.877 [] DEBUG io.camunda.zeebe.util.actor - Closing actor thread ground '-zb-fs-workers'
08:55:51.878 [] DEBUG io.camunda.zeebe.util.actor - Closing actor thread ground '-zb-fs-workers': closed successfully
08:55:51.878 [] DEBUG io.camunda.zeebe.util.actor - Closing actor thread ground '-zb-actors'
08:55:51.879 [] DEBUG io.camunda.zeebe.util.actor - Closing actor thread ground '-zb-actors': closed successfully

java.lang.AssertionError: 
Expecting:
  [(PROCESS, ACTIVATE_ELEMENT),
    (PROCESS, ELEMENT_ACTIVATING),
    (PROCESS, ELEMENT_ACTIVATED),
    (START_EVENT, ACTIVATE_ELEMENT),
    (START_EVENT, ELEMENT_ACTIVATING),
    (START_EVENT, ELEMENT_ACTIVATED),
    (START_EVENT, COMPLETE_ELEMENT),
    (START_EVENT, ELEMENT_COMPLETING),
    (START_EVENT, ELEMENT_COMPLETED),
    (SEQUENCE_FLOW, SEQUENCE_FLOW_TAKEN),
    (SERVICE_TASK, ACTIVATE_ELEMENT),
    (SERVICE_TASK, ELEMENT_ACTIVATING),
    (SERVICE_TASK, ELEMENT_ACTIVATED),
    (SERVICE_TASK, TERMINATE_ELEMENT),
    (SERVICE_TASK, TERMINATE_ELEMENT),
    (SERVICE_TASK, ELEMENT_TERMINATING),
    (SERVICE_TASK, ELEMENT_TERMINATED),
    (BOUNDARY_EVENT, ELEMENT_ACTIVATING),
    (BOUNDARY_EVENT, ELEMENT_ACTIVATED),
    (BOUNDARY_EVENT, COMPLETE_ELEMENT),
    (SERVICE_TASK, TERMINATE_ELEMENT),
    (BOUNDARY_EVENT, COMPLETE_ELEMENT)]
to contain subsequence:
  [(EVENT_SUB_PROCESS, ELEMENT_ACTIVATED),
    (EVENT_SUB_PROCESS, ELEMENT_COMPLETED),
    (PROCESS, ELEMENT_COMPLETED)]


	at io.camunda.zeebe.engine.processing.bpmn.subprocess.InterruptingEventSubprocessConcurrencyTest.shouldInterruptBoundaryEvent(InterruptingEventSubprocessConcurrencyTest.java:174)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke0(Native Method)
	at java.base/jdk.internal.reflect.NativeMethodAccessorImpl.invoke(NativeMethodAccessorImpl.java:62)
	at java.base/jdk.internal.reflect.DelegatingMethodAccessorImpl.invoke(DelegatingMethodAccessorImpl.java:43)
	at java.base/java.lang.reflect.Method.invoke(Method.java:566)
	at org.junit.runners.model.FrameworkMethod$1.runReflectiveCall(FrameworkMethod.java:59)
	at org.junit.internal.runners.model.ReflectiveCallable.run(ReflectiveCallable.java:12)
	at org.junit.runners.model.FrameworkMethod.invokeExplosively(FrameworkMethod.java:56)
	at org.junit.internal.runners.statements.InvokeMethod.evaluate(InvokeMethod.java:17)
	at org.junit.rules.TestWatcher$1.evaluate(TestWatcher.java:61)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.ExternalResource$1.evaluate(ExternalResource.java:54)
	at org.junit.rules.RunRules.evaluate(RunRules.java:20)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.BlockJUnit4ClassRunner$1.evaluate(BlockJUnit4ClassRunner.java:100)
	at org.junit.runners.ParentRunner.runLeaf(ParentRunner.java:366)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:103)
	at org.junit.runners.BlockJUnit4ClassRunner.runChild(BlockJUnit4ClassRunner.java:63)
	at org.junit.runners.ParentRunner$4.run(ParentRunner.java:331)
	at org.junit.runners.ParentRunner$1.schedule(ParentRunner.java:79)
	at org.junit.runners.ParentRunner.runChildren(ParentRunner.java:329)
	at org.junit.runners.ParentRunner.access$100(ParentRunner.java:66)
	at org.junit.runners.ParentRunner$2.evaluate(ParentRunner.java:293)
	at org.junit.runners.ParentRunner$3.evaluate(ParentRunner.java:306)
	at org.junit.runners.ParentRunner.run(ParentRunner.java:413)
	at org.junit.runner.JUnitCore.run(JUnitCore.java:137)
	at com.intellij.junit4.JUnit4IdeaTestRunner.startRunnerWithArgs(JUnit4IdeaTestRunner.java:69)
	at com.intellij.rt.junit.IdeaTestRunner$Repeater.startRunnerWithArgs(IdeaTestRunner.java:33)
	at com.intellij.rt.junit.JUnitStarter.prepareStreamsAndStart(JUnitStarter.java:220)
	at com.intellij.rt.junit.JUnitStarter.main(JUnitStarter.java:53)


Process finished with exit code 255


```

--- a/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/subprocess/InterruptingEventSubprocessConcurrencyTest.java
+++ b/engine/src/test/java/io/camunda/zeebe/engine/processing/bpmn/subprocess/InterruptingEventSubprocessConcurrencyTest.java
@@ -15,7 +15,9 @@ import io.camunda.zeebe.engine.util.RecordToWrite;
 import io.camunda.zeebe.model.bpmn.Bpmn;
 import io.camunda.zeebe.model.bpmn.BpmnModelInstance;
 import io.camunda.zeebe.model.bpmn.builder.ProcessBuilder;
+import io.camunda.zeebe.protocol.impl.record.value.message.MessageRecord;
 import io.camunda.zeebe.protocol.record.intent.IncidentIntent;
+import io.camunda.zeebe.protocol.record.intent.MessageIntent;
 import io.camunda.zeebe.protocol.record.intent.MessageSubscriptionIntent;
 import io.camunda.zeebe.protocol.record.intent.ProcessInstanceIntent;
 import io.camunda.zeebe.protocol.record.intent.ProcessMessageSubscriptionIntent;
@@ -208,6 +210,67 @@ public class InterruptingEventSubprocessConcurrencyTest {
         .withCorrelationKey("123")
         .withVariables(Map.of("key", "123"))
         .publish();
+
+    // then
+    assertThat(
+            RecordingExporter.processInstanceRecords()
+                .withProcessInstanceKey(processInstanceKey)
+                .limitToProcessInstanceCompleted())
+        .extracting(r -> tuple(r.getValue().getBpmnElementType(), r.getIntent()))
+        .containsSubsequence(
+            tuple(BpmnElementType.EVENT_SUB_PROCESS, ProcessInstanceIntent.ELEMENT_ACTIVATED),
+            tuple(BpmnElementType.EVENT_SUB_PROCESS, ProcessInstanceIntent.ELEMENT_COMPLETED),
+            tuple(BpmnElementType.PROCESS, ProcessInstanceIntent.ELEMENT_COMPLETED));
+  }
+
+  @Test
+  // https://github.com/camunda-cloud/zeebe/issues/6587
+  public void shouldInterruptBoundaryEvent() {
+    // given
+    final ProcessBuilder process = Bpmn.createExecutableProcess(PROCESS_ID);
+
+    process
+        .eventSubProcess("event_sub_proc")
+        .startEvent("event_sub_start")
+        .interrupting(true)
+        .message(b -> b.name(MSG_NAME).zeebeCorrelationKeyExpression("key"))
+        .endEvent("event_sub_end");
+
+    final BpmnModelInstance model =
+        process
+            .startEvent("start_proc")
+            .sequenceFlowId("toParallel")
+            .serviceTask("serviceTask", s -> s.zeebeJobType("task"))
+            .boundaryEvent("boundary")
+            .cancelActivity(true)
+            .message(m -> m.name("msg").zeebeCorrelationKeyExpression("key"))
+            .endEvent("end_proc")
+            .done();
+
+    engineRule.deployment().withXmlResource(model).deploy();
+
+    final long processInstanceKey =
+        engineRule
+            .processInstance()
+            .ofBpmnProcessId(PROCESS_ID)
+            .withVariables(Map.of("key", 123))
+            .create();
+
+    // when
+    RecordingExporter.messageSubscriptionRecords(MessageSubscriptionIntent.CREATED)
+        .withProcessInstanceKey(processInstanceKey)
+        .withMessageName("msg")
+        .await();
+
+    engineRule.writeRecords(
+        RecordToWrite.command()
+            .message(
+                MessageIntent.PUBLISH,
+                new MessageRecord().setName("msg").setCorrelationKey("123").setTimeToLive(0)),
+        RecordToWrite.command()
+            .message(
+                MessageIntent.PUBLISH,
+                new MessageRecord().setName(MSG_NAME).setCorrelationKey("123").setTimeToLive(0)));
 
     // then
     assertThat(

